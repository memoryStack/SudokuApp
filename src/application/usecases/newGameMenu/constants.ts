// TODO: how to compose enums so that's it's easier to edit or re-organize them
// TODO: these names are confusing, fix these
export enum AUTO_GENERATED_NEW_GAME_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
}

export enum NEW_GAME_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
    CUSTOMIZE_PUZZLE = 'CUSTOMIZE_PUZZLE',
}

export enum START_GAME_MENU_ITEMS_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
    CUSTOMIZE_PUZZLE = 'CUSTOMIZE_PUZZLE',
    RESUME = 'RESUME'
}
