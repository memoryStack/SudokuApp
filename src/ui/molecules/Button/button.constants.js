const BUTTON_TYPES = {
    FILLED: 'filled',
    TONAL: 'tonal',
    TEXT: 'text',
    /*
        elevated
        contained
    */
}

const BUTTON_STATES = {
    ENABLED: 'enabled',
    DISABLED: 'disabled',
}

// TODO: not getting used anywhere, look for use-cases and add tokens
//          for sizes as well
const BUTTON_SIZES = {
    SMALL: 'small',
    MEDIUM: 'medium',
    LARGE: 'large',
}

export {
    BUTTON_STATES,
    BUTTON_TYPES,
    BUTTON_SIZES,
}
