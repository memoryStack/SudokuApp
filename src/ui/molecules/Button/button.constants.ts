export enum BUTTON_TYPES {
    FILLED = 'filled',
    TONAL = 'tonal',
    TEXT = 'text',
    /*
        elevated
        contained
    */
}

export enum BUTTON_STATES {
    ENABLED = 'enabled',
    DISABLED = 'disabled',
}

// TODO: not getting used anywhere, look for use-cases and add tokens
//          for sizes as well
export enum BUTTON_SIZES {
    SMALL = 'small',
    MEDIUM = 'medium',
    LARGE = 'large',
}
