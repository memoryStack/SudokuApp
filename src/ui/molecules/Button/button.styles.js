import { StyleSheet } from 'react-native'

import _get from '@lodash/get'

import { hexToRGBA } from '@utils/util'

// Note: can't use opacity property here because it will dim down text as well
//      looks like opacity dim downs whole container's contents along with container
//      so work-around is to use rgba() function for showing disabled states

const getStyles = ({
    type, state, isIconAvailable, isLabelAvailable,
}, theme) => {
    const buttonConfigs = {
        type, state, isIconAvailable, isLabelAvailable,
    }

    return StyleSheet.create({
        defaultContainer: {
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: getContainerBackgroundColor(buttonConfigs, theme),
            height: _get(theme, ['button', type, 'layout', 'container', 'height']),
            borderRadius: _get(theme, ['button', type, 'layout', 'container', 'border-radius']),
            ...getContainerPaddings(buttonConfigs, theme),
        },
        defaultText: {
            color: getTextColor(buttonConfigs, theme),
            ..._get(theme, ['button', type, 'layout', 'label-text', 'font']),
        },
        icon: { // TODO: how these properties will work with Svg icons. test that
            color: getIconColor(buttonConfigs, theme),
            width: _get(theme, ['button', type, 'layout', 'icon', 'size']),
            height: _get(theme, ['button', type, 'layout', 'icon', 'size']),
        },
    })
}

const getContainerBackgroundColor = ({ type, state }, theme) => {
    const { background = 'transparent', opacity } = _get(theme, ['button', type, 'color', 'container', state], {})
    return hexToRGBA(background, opacity)
}

const getContainerPaddings = ({
    type, isIconAvailable, isLabelAvailable,
}, theme) => {
    // TODO: what will happen if only icon is available ??
    //      it will need some refactoring
    let paddingLeft; let paddingRight
    if (isIconAvailable && isLabelAvailable) {
        paddingLeft = _get(theme, ['button', type, 'layout', 'container', 'padding', 'with-icon', 'icon-side-padding'])
        paddingRight = _get(theme, ['button', type, 'layout', 'container', 'padding', 'with-icon', 'icon-opposite-side-padding'])
    } else if (isLabelAvailable) {
        paddingLeft = _get(theme, ['button', type, 'layout', 'container', 'padding', 'without-icon', 'padding-left'])
        paddingRight = _get(theme, ['button', type, 'layout', 'container', 'padding', 'without-icon', 'padding-right'])
    }
    return { paddingLeft, paddingRight }
}

const getTextColor = ({ type, state }, theme) => {
    const { color, opacity } = _get(theme, ['button', type, 'color', 'label-text', state], {})
    return hexToRGBA(color, opacity)
}

const getIconColor = ({ type, state }, theme) => {
    const { color, opacity } = _get(theme, ['button', type, 'color', 'icon', state], {})
    return hexToRGBA(color, opacity)
}

export { getStyles }
