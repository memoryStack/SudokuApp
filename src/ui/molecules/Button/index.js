export { default } from './Button'
export * from './button.constants'
