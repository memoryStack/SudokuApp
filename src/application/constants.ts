import { AUTO_GENERATED_NEW_GAME_IDS } from "./usecases/newGameMenu/constants"

export enum BOARD_MOVES_TYPES {
    ADD = 'ADD',
    REMOVE = 'REMOVE',
}

export enum PENCIL_STATE {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
}

export enum GAME_STATE {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
    DISPLAY_HINT = 'DISPLAY_HINT',
    GAME_SELECT = 'GAME_SELECT',
    OVER_SOLVED = 'SOLVED',
    OVER_UNSOLVED = 'UNSOLVED',
}

export enum LEVEL_DIFFICULTIES {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
}

export const GAME_LEVEL_VS_CLUES = {
    [AUTO_GENERATED_NEW_GAME_IDS.EASY]: 36,
    [AUTO_GENERATED_NEW_GAME_IDS.MEDIUM]: 32,
    [AUTO_GENERATED_NEW_GAME_IDS.HARD]: 28,
    [AUTO_GENERATED_NEW_GAME_IDS.EXPERT]: 24,
}
