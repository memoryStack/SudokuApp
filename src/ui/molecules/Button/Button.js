import React from 'react'

import PropTypes from 'prop-types'

import _noop from '@lodash/noop'

import { useStyles } from '@utils/customHooks/useStyles'

import { Touchable, TouchableTypes } from 'src/apps/components/Touchable'

import Text from '@ui/atoms/Text'

import { getStyles } from './button.styles'
import { BUTTON_STATES, BUTTON_TYPES, HIT_SLOP } from './button.constants'

const Button_ = ({
    onClick,
    containerStyle = null,
    label = '',
    textStyles = null,
    avoidDefaultContainerStyles = false,
    type,
    state,
    ...rest
}) => {
    const styles = useStyles(getStyles, {
        type,
        state,
        isLabelAvailable: !!label,
        isIconAvailable: false,
    })

    return (
        <Touchable
            touchable={TouchableTypes.opacity}
            style={[avoidDefaultContainerStyles ? null : styles.defaultContainer, containerStyle]}
            avoidDefaultStyles={avoidDefaultContainerStyles}
            onPress={onClick}
            disabled={state === BUTTON_STATES.DISABLED}
            hitSlop={type === BUTTON_TYPES.TEXT ? HIT_SLOP : null}
            accessibilityRole="button"
            {...rest}
        >
            <Text style={{ ...styles.defaultText, ...textStyles }}>
                {label}
            </Text>
        </Touchable>
    )
}

export default React.memo(Button_)

Button_.propTypes = {
    onClick: PropTypes.func,
    containerStyle: PropTypes.object,
    label: PropTypes.string,
    textStyles: PropTypes.object,
    avoidDefaultContainerStyles: PropTypes.bool,
    type: PropTypes.oneOf(Object.values(BUTTON_TYPES)),
    state: PropTypes.oneOf(Object.values(BUTTON_STATES)),
}

Button_.defaultProps = {
    onClick: _noop,
    containerStyle: null,
    label: '',
    textStyles: null,
    avoidDefaultContainerStyles: false,
    type: BUTTON_TYPES.FILLED,
    state: BUTTON_STATES.ENABLED,
}
