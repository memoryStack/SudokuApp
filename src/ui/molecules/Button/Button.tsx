import React from 'react'

import { StyleProp, ViewStyle } from 'react-native'

import _noop from '@lodash/noop'

import Text, { TEXT_VARIATIONS } from '@ui/atoms/Text'

import { useStyles } from '@utils/customHooks/useStyles'

import { useTranslation } from '../../../i18n/hooks/useTranslation'
import { Touchable } from '../../../apps/components/Touchable'

import { getStyles } from './button.styles'
import { BUTTON_STATES, BUTTON_TYPES } from './button.constants'

interface Props {
    children?: React.ReactNode,
    onPress: () => unknown,
    containerStyle?: StyleProp<ViewStyle>,
    label?: string,
    textStyles?: StyleProp<ViewStyle>,
    avoidDefaultContainerStyles?: boolean,
    type?: BUTTON_TYPES,
    state?: BUTTON_STATES,
    textType?: TEXT_VARIATIONS,
    testID?: string,
}

const Button_: React.FC<Props> = ({
    onPress = _noop,
    containerStyle = null,
    label = '',
    textStyles = {},
    avoidDefaultContainerStyles = false,
    type = BUTTON_TYPES.FILLED,
    state = BUTTON_STATES.ENABLED,
    children = null,
    textType,
    ...rest
}) => {
    const { t } = useTranslation()

    const styles = useStyles(getStyles, {
        type,
        state,
        isLabelAvailable: !!label,
        isIconAvailable: false,
    })

    const getTextFinalStyles = () => Object.assign({}, styles.labelDefaultColor, !textType && styles.labelDefaultFont, textStyles)

    const renderLabel = () => (
        <Text style={getTextFinalStyles()} type={textType}>
            {t(label)}
        </Text>
    )

    return (
        <Touchable
            style={[avoidDefaultContainerStyles ? null : styles.defaultContainer, containerStyle]}
            avoidDefaultStyles={avoidDefaultContainerStyles}
            onPress={onPress}
            disabled={state === BUTTON_STATES.DISABLED}
            addHitSlop={type === BUTTON_TYPES.TEXT}
            accessibilityRole="button"
            {...rest}
        >
            {children || renderLabel()}
        </Touchable>
    )
}

export default React.memo(Button_)
