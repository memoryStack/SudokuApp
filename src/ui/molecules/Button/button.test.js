import * as React from 'react'

import { render, screen, fireEvent } from 'testing-utils'
import { PersistentRender } from '@utils/testing/persistentRender'

import Button from './Button'
import { BUTTON_STATES, BUTTON_TYPES } from './button.constants'

test('snapshot for filled button in enabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.FILLED}
            state={BUTTON_STATES.ENABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for filled button in disabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.FILLED}
            state={BUTTON_STATES.DISABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for tonal button in enabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.TONAL}
            state={BUTTON_STATES.ENABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for tonal button in disabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.TONAL}
            state={BUTTON_STATES.DISABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for text button in enabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.TEXT}
            state={BUTTON_STATES.ENABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for text button in disabled state', () => {
    const tree = render(
        <Button
            label="test button"
            type={BUTTON_TYPES.TEXT}
            state={BUTTON_STATES.DISABLED}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

// should i add test-cases if prop-override colors are working fine or not ?
describe('Button functionality', () => {
    test('renders only the outer circle if not selected', () => {
        const buttonLabel = 'test button'
        render(<Button label={buttonLabel} />)

        expect(screen.getByText(buttonLabel)).toBeTruthy()
    })

    test('performs onClick when pressed', () => {
        const buttonLabel = 'test button'
        const onClick = jest.fn()
        render(<Button label={buttonLabel} onClick={onClick} />)

        fireEvent.press(screen.getByText(buttonLabel))
        expect(onClick).toHaveBeenCalledTimes(1)
    })

    test('for disabled state onClick is not called on user press', () => {
        const buttonLabel = 'test button'
        const onClick = jest.fn()
        render(<Button state={BUTTON_STATES.DISABLED} label={buttonLabel} onClick={onClick} />)

        fireEvent.press(screen.getByText(buttonLabel))
        expect(onClick).toHaveBeenCalledTimes(0)
    })

    test.only('changes appearence when button state is toggled', () => {
        const buttonLabel = 'test button'

        const aPersistentRender = new PersistentRender({
            getRenderingResultToCompare: () => {
                const element = screen.getByText(buttonLabel)
                return [element.props.style.color]
            },
        })
        aPersistentRender.render(<Button state={BUTTON_STATES.ENABLED} label={buttonLabel} />)
        aPersistentRender.update(<Button state={BUTTON_STATES.DISABLED} label={buttonLabel} />)

        expect(aPersistentRender.getChangedRenderingResultStatus()).toEqual([true])
    })
})
