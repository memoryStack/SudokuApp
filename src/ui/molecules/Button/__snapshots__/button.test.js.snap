// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for filled button in disabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(27, 23, 28, 0.12)",
      "borderRadius": 20,
      "display": "flex",
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": 24,
      "paddingRight": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(27, 23, 28, 0.38)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;

exports[`snapshot for filled button in enabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(103, 80, 164, 1)",
      "borderRadius": 20,
      "display": "flex",
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": 24,
      "paddingRight": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(255, 255, 255, 1)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;

exports[`snapshot for text button in disabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 16,
      "left": 16,
      "right": 16,
      "top": 16,
    }
  }
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderRadius": undefined,
      "display": "flex",
      "height": undefined,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": undefined,
      "paddingRight": undefined,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(27, 23, 28, 0.38)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;

exports[`snapshot for text button in enabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 16,
      "left": 16,
      "right": 16,
      "top": 16,
    }
  }
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderRadius": undefined,
      "display": "flex",
      "height": undefined,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": undefined,
      "paddingRight": undefined,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(103, 80, 164, 1)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;

exports[`snapshot for tonal button in disabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(27, 23, 28, 0.12)",
      "borderRadius": 20,
      "display": "flex",
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": 24,
      "paddingRight": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(27, 23, 28, 0.38)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;

exports[`snapshot for tonal button in enabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(232, 222, 248, 1)",
      "borderRadius": 20,
      "display": "flex",
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "paddingLeft": 24,
      "paddingRight": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(29, 25, 43, 1)",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 16,
        "fontWeight": "500",
        "includeFontPadding": false,
        "letterSpacing": 0.1,
        "lineHeight": 24,
        "textAlignVertical": "center",
      }
    }
  >
    test button
  </Text>
</View>
`;
